clients/apps/web/src/components/CustomerPortal/OrderPaymentRetry.tsx

```diff
@@ -0,0 +1,406 @@
+import Button from '@polar-sh/ui/components/atoms/Button'
+import { PaymentElement } from '@stripe/react-stripe-js'
+import {
+  ConfirmationToken,
+  Stripe,
+  StripeElements,
+  StripeError,
+} from '@stripe/stripe-js'
+import { useEffect, useRef, useState } from 'react'
+
+interface OrderPaymentRetryProps {
+  order: schemas['CustomerOrder']
+  stripe: Stripe | null
+  elements: StripeElements | null
+  api: Client
+  onSuccess: () => void
+  onError: (error: string) => void
+  onClose: () => void
+}
+
+export const OrderPaymentRetry = ({
+  order,
+  stripe,
+  elements,
+  api,
+  onSuccess,
+  onError,
+  onClose,
+}: OrderPaymentRetryProps) => {
+  const confirmOrderPayment = useCustomerOrderConfirmPayment(api)
+  const checkPaymentStatus = useCustomerOrderPaymentStatus(api)
+  const [isProcessing, setIsProcessing] = useState(false)
+  const [isPolling, setIsPolling] = useState(false)
+  const [hasSubmitted, setHasSubmitted] = useState(false)
+  const [paymentComplete, setPaymentComplete] = useState(false)
+  const [showRetryButton, setShowRetryButton] = useState(false)
+
+  const pollingIntervalRef = useRef<NodeJS.Timeout | null>(null)
+
+  const cleanupPolling = () => {
+    if (pollingIntervalRef.current) {
+      clearInterval(pollingIntervalRef.current)
+      pollingIntervalRef.current = null
+    }
+  }
+
+  const resetProcessingStates = () => {
+    setIsProcessing(false)
+    setIsPolling(false)
+    setHasSubmitted(false)
+    setPaymentComplete(false)
+    setShowRetryButton(false)
+    cleanupPolling()
+  }
+
+  const handlePaymentCompletion = (success: boolean, message?: string) => {
+    setIsProcessing(false)
+    setIsPolling(false)
+    setPaymentComplete(true)
+
+    if (success) {
+      onSuccess()
+    } else {
+      setShowRetryButton(true)
+      if (message) {
+        onError(message)
+      }
+    }
+  }
+
+  const handlePaymentStatus = async (result: any) => {
+    switch (result.status) {
+      case 'succeeded':
+        handlePaymentCompletion(true)
+        break
+
+      case 'requires_action':
+        await handlePaymentAction(result)
+        break
+
+      default:
+        handlePaymentCompletion(
+          false,
+          result.error || 'Payment failed, please try again.',
+        )
+    }
+  }
+
+  const handlePaymentAction = async (result: any) => {
+    if (!stripe) {
+      handlePaymentCompletion(
+        false,
+        'Stripe instance is required for payment actions',
+      )
+      return
+    }
+
+    if (!result.client_secret) {
+      handlePaymentCompletion(
+        false,
+        'Payment requires additional authentication',
+      )
+      return
+    }
+
+    const { error } = await stripe.handleNextAction({
+      clientSecret: result.client_secret,
+    })
+
+    if (error) {
+      handlePaymentCompletion(
+        false,
+        error.message || 'Payment authentication failed',
+      )
+    } else {
+      pollForWebhookResult()
```

Comment:
Naive: since we process the payment almost synchronously now, couldn't we just wait like ~1 sec and then refetch the order, call it a day? 😄

Reply:
When status is failed or succeeded, we show that to the user immediately. If the status is `required_action` we do the pulling.

<img width="3840" height="3288" alt="Untitled diagram _ Mermaid Chart-2025-08-19-083029" src="https://github.com/user-attachments/assets/9eec77a5-5604-4afa-8206-50aded83d1fa" />

Reply:
Yes, my point was more: do we actually need to do a polling? Most likely, the order will be marked as paid shortly after a successful payment, so we could just refetch the latest order data.